docusaurus/docs/how-to-guides/panel-plugins/options-migration-handler.md

```diff
@@ -0,0 +1,157 @@
+import { SimplePanel } from './components/SimplePanel';
+import { SimpleOptions } from './types';
+import { migrationHandler } from './migrationHandler';
+
+export const plugin = new PanelPlugin<SimpleOptions>(SimplePanel)
+  .setPanelOptions((builder) => {
+    // ... panel options configuration
+  })
+  // define the migration handler
+  .setMigrationHandler(migrationHandler);
+```
+
+:::note
+
+The migration handler is only called when the installed plugin version differs from the version used to generate a panel. Changes made to the panel are not automatically persisted, users need to manually save the dashboard after a panel is migrated.
+:::
+
+## Implement a migration handler
+
+The migration handler function receives the entire panel model as an argument and should return the updated panel options. Here's a basic structure:
+
+```ts title="migrationHandler.ts"
+import { PanelModel } from '@grafana/data';
+import { SimpleOptions } from './types';
+
+function migrationHandler(panel: PanelModel<Partial<SimpleOptions>>) {
+  // panel.options contains the current panel options stored in the dashboard
+  const options = Object.assign({}, panel.options);
+
+  // Perform option migrations here
+
+  return options;
+}
+```
+
+## Common migration scenarios
+
+### Handling new options
+
+When adding new panel configuration options, set default values for them:
+
+```ts
+if (options.newFeature === undefined) {
+  panel.options.newFeature = 'defaultValue';
+}
+```
+
+### Renaming options
+
+If you've renamed an option, transfer the value from the old option to the new one:
+
+```ts
+if (panel.options.oldOptionName) {
+  panel.options.newOptionName = panel.options.oldOptionName;
+  // make sure to remove the old option
+  delete panel.options.oldOptionName;
+}
+```
+
+### Adjusting for changed options
+
+When removing valid options or changing valid selections, set safe defaults:
+
+```ts
+const validOptions = ['option1', 'option2', 'option3'];
+if (!validOptions.includes(panel.options.someOption)) {
+  panel.options.someOption = validOptions[0];
+}
+```
+
+Or migrate the existing value to the new option:
+
+```ts
+// e.g. the options of displayType were renamed
+// from bar, line, pie to barChart, linePlot and pieChart
+if (options.displayType) {
+  switch (options.displayType) {
+    case 'bar':
+      options.displayType = 'barChart';
+      break;
+    case 'line':
+      options.displayType = 'linePlot';
+      break;
+    case 'pie':
+      options.displayType = 'pieChart';
+      break;
+  }
+}
+```
+
+### Version-specific adjustments
+
+You might want to run specific adjustments based on the previously installed plugin version.
+
+you can use the `pluginVersion` property. This property:
+
+- is empty if it is the first time a migration handler is ever called. e.g. the plugin didn't implement a migration handler before.
+- is set to the version of the plugin when the panel was last saved after a migration handler was called.
```

Comment:
```suggestion
You might want to base migration decissions based on the version that was used to write a panel. For this, you can use the `pluginVersion` property. This property is empty the first time the migration handler is ever used but after that it will be set to the plugin version used to save the panel.

For example, imagine the following history for a plugin:

 - In v1, the plugin didn't have any migration code.
 - In v2, the plugin introduced the first migration code.
 - In v3, the plugin changed again with new migration steps.
 
In that scenario, the migration handler would look like this:

```ts title="migrationHandler.ts"
import { config } from '@grafana/runtime';

function migrationHandler(panel: PanelModel<SimpleOptions>) {
  const options = Object.assign({}, panel.options);
  const pluginVersion = panel?.pluginVersion ?? '';

  if (pluginVersion === '') {
    // Plugin version was v1.x
    // Needs logic to migrate v1 -> v3
    options.displayMode = 'compact';
    options.displayType = 'linePlot';
  }
  
  if (pluginVersion.startsWith('2.') {
    // Panel was last saved with version v2.x
    // Needs logic to migrate v2 -> v3
    options.displayMode = 'compact';
  }

  return options;
}
```

Reply:
@josmperez What do you think?